### PR TITLE
Add UniqueIds for improved scalable UUID generation

### PR DESCRIPTION
**Pull request by schlosna**
## Before this PR

`UUID.randomUUID()` generation and string serialization can be a scalability bottleneck, especially under high volume AtlasDB workloads. The top two I have seen recently pop up in profiles:

- `com.palantir.atlasdb.transaction.impl.TransactionRetryStrategy.runWithRetry(IntPredicate, TransactionRetryStrategy$Retryable):71`
  - https://github.com/palantir/atlasdb/blob/1acfd3d41b3e7e504320ea76b40ee9c134dab91a/atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategy.java#L71
- `com.palantir.lock.client.LeasedLockToken.of(ConjureLockToken, Lease):39`
  - https://github.com/palantir/atlasdb/blob/1acfd3d41b3e7e504320ea76b40ee9c134dab91a/lock-api/src/main/java/com/palantir/lock/client/LeasedLockToken.java#L39

image

There is some OpenJDK work either completed or in flight that is worth tracking:

- [JDK-8308804 Improve UUID.randomUUID performance with bulk/scalable PRNG access](https://bugs.openjdk.org/browse/JDK-8308804) , [PR 14135](https://github.com/openjdk/jdk/pull/14135) (not yet integrated, closed as stale)
- [JDK-8311207: Cleanup for Optimization for UUID.toString](https://bugs.openjdk.org/browse/JDK-8311207) , [PR 14745](https://github.com/openjdk/jdk/pull/14745)
- [JDK-8310502 : Optimization for j.l.Long.fastUUID()](https://bugs.openjdk.org/browse/JDK-8310502) , [PR 14578](https://github.com/openjdk/jdk/pull/14578)

Also worth tracking and possibly utilizing one of the newer proposed UUID formats:

- https://uuid6.github.io/uuid6-ietf-draft/
- https://datatracker.ietf.org/doc/draft-peabody-dispatch-new-uuid-format/
- https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html#v8

## After this PR
Moving reusable UniqueIds from elsewhere to `tritium-ids` for use in  https://github.com/palantir/atlasdb/pull/6726 and other places.

==COMMIT_MSG==
Add UniqueIds for improved scalable UUID generation
==COMMIT_MSG==

```
16x Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz cores
VM version: JDK 17.0.8, OpenJDK 64-Bit Server VM, 17.0.8+7-LTS
Benchmark                             Mode  Cnt    Score     Error   Units
UuidBenchmark.jdkRandomUuidMax       thrpt    3    1.673 ?   1.378  ops/us // negative scale!
UuidBenchmark.jdkRandomUuidOne       thrpt    3    2.511 ?   0.272  ops/us
UuidBenchmark.jugSecureRandomMax     thrpt    3    2.008 ?   1.307  ops/us
UuidBenchmark.pseudoRandomUuidV4Max  thrpt    3  236.455 ? 106.688  ops/us
UuidBenchmark.pseudoRandomUuidV4One  thrpt    3   29.096 ?   3.265  ops/us
UuidBenchmark.randomUuidV4Max        thrpt    3   28.202 ?   8.763  ops/us
UuidBenchmark.randomUuidV4One        thrpt    3    3.388 ?   0.266  ops/us
UuidBenchmark.secureRandomMax        thrpt    3    2.015 ?   0.923  ops/us
UuidBenchmark.threadLocalMax         thrpt    3    1.652 ?   1.700  ops/us
UuidBenchmark.threadLocalSha1Max     thrpt    3   27.365 ?   5.918  ops/us
UuidBenchmark.twoLongsMax            thrpt    3    0.676 ?   0.450  ops/us
```

```
MacBook Pro 2021 10x aarch64 cores
VM version: JDK 17.0.8.1, OpenJDK 64-Bit Server VM, 17.0.8.1+8-LTS
Benchmark                             Mode  Cnt    Score    Error   Units
UuidBenchmark.jdkRandomUuidMax       thrpt    3    2.263 ?  0.358  ops/us // negative scale!
UuidBenchmark.jdkRandomUuidOne       thrpt    3    4.005 ?  0.211  ops/us
UuidBenchmark.jugSecureRandomMax     thrpt    3    2.902 ?  0.156  ops/us
UuidBenchmark.pseudoRandomUuidV4Max  thrpt    3  284.048 ? 48.930  ops/us
UuidBenchmark.pseudoRandomUuidV4One  thrpt    3   38.915 ? 52.533  ops/us
UuidBenchmark.randomUuidV4Max        thrpt    3   31.332 ?  4.457  ops/us
UuidBenchmark.randomUuidV4One        thrpt    3    4.583 ?  0.705  ops/us
UuidBenchmark.secureRandomMax        thrpt    3    2.934 ?  0.115  ops/us
UuidBenchmark.threadLocalMax         thrpt    3    2.296 ?  0.223  ops/us
UuidBenchmark.threadLocalSha1Max     thrpt    3   26.574 ? 12.350  ops/us
UuidBenchmark.twoLongsMax            thrpt    3    1.122 ?  0.405  ops/us
```

### CODE REVIEW

File: tritium-ids/src/main/java/com/palantir/tritium/ids/UniqueIds.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            log.warn("Falling back to default SecureRandom", e);
+            return new SecureRandom();
+        }
+    });
+
+    /**
+     * Returns a unique {@link UUID} using.
+     */
+    public static UUID v4PseudoRandomUuid() {
+        return v4RandomUuid(ThreadLocalRandom.current());
```
Comment (schlosna) [inline tritium-ids/src/main/java/com/palantir/tritium/ids/UniqueIds.java]: On the fence whether to include this, leaning toward removing.

File: tritium-ids/src/main/java/com/palantir/tritium/ids/UniqueIds.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    /**
+     * Returns a unique {@link UUID}.
+     */
+    public static UUID v4RandomUuid() {
```
Comment (schlosna) [inline tritium-ids/src/main/java/com/palantir/tritium/ids/UniqueIds.java]: open to thoughts on naming here ( 🟦 🚲 🏠 ). I haven't added the draft v6, v7, or v8 UUID types but want space in API for those.

File: tritium-ids/src/main/java/com/palantir/tritium/ids/UniqueIds.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        }
+
+        return new UUID(mostSigBits, leastSigBits);
+    }
+
+    private static byte[] bytes(Random rand) {
+        byte[] data = new byte[16];
+        rand.nextBytes(data);
+        return data;
+    }
```
Comment (carterkozak) [inline tritium-ids/src/main/java/com/palantir/tritium/ids/UniqueIds.java:86]: I wonder if it's reasonable to use a pair of `random.nextLong()` results?

Comment (schlosna) [reply tritium-ids/src/main/java/com/palantir/tritium/ids/UniqueIds.java:86]: I can benchmark that.

My sense is that the contention is actually on the SecureRandom instance method invocations and not the underlying entropy source, so the two calls would contend more for the cases where there is a single shared `SecureRandom`.

I will also try out per processor pool of random bytes.

Comment (schlosna) [reply tritium-ids/src/main/java/com/palantir/tritium/ids/UniqueIds.java:86]: I tested out a few iterations of pooled byte arrays in [ds/ids-pool-chunks](https://github.com/palantir/tritium/compare/ds/ids-pool-chunks?expand=1) and tweaking the `CHUNKS_TO_ALLOCATE`, but `ThreadLocal<SecureRandom>` for `randomUuidV4` seems worth it for near linear core scaling in `randomUuidV4Max`, compared to the negative scaling due to contention in `jdkRandomUuidMax`.

We may want to reassess post-JDK21 + virtual threads as the cost of `ThreadLocal` will change, and we may want to just pool `SecureRandom` instances.
> I wonder if it's reasonable to use a pair of `random.nextLong()` results?

The `UuidBenchmark.twoLongsMax` above represents this case

### `ThreadLocal<SecureRandom>` (this branch)

```
16x Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz cores
VM version: JDK 17.0.8, OpenJDK 64-Bit Server VM, 17.0.8+7-LTS
Benchmark                          Mode  Cnt   Score   Error   Units
UuidBenchmark.jdkRandomUuidMax    thrpt    3   1.689 ? 1.397  ops/us
UuidBenchmark.jdkRandomUuidOne    thrpt    3   2.488 ? 0.252  ops/us
UuidBenchmark.jugSecureRandomMax  thrpt    3   2.044 ? 1.434  ops/us
UuidBenchmark.secureRandomMax     thrpt    3   2.053 ? 1.473  ops/us
UuidBenchmark.threadLocalMax      thrpt    3  28.340 ? 2.180  ops/us
UuidBenchmark.twoLongsMax         thrpt    3   0.699 ? 1.036  ops/us
UuidBenchmark.randomUuidV4Max     thrpt    3  27.893 ? 2.183  ops/us
UuidBenchmark.randomUuidV4One     thrpt    3   3.393 ? 0.312  ops/us
```

```
MacBook Pro 2021 10x aarch64 cores
VM version: JDK 17.0.8.1, OpenJDK 64-Bit Server VM, 17.0.8.1+8-LTS
Benchmark                          Mode  Cnt   Score   Error   Units
UuidBenchmark.jdkRandomUuidMax    thrpt    3   2.265 ? 0.233  ops/us
UuidBenchmark.jdkRandomUuidOne    thrpt    3   3.864 ? 0.064  ops/us
UuidBenchmark.jugSecureRandomMax  thrpt    3   2.886 ? 0.186  ops/us
UuidBenchmark.secureRandomMax     thrpt    3   2.847 ? 0.851  ops/us
UuidBenchmark.threadLocalMax      thrpt    3  26.719 ? 5.500  ops/us
UuidBenchmark.twoLongsMax         thrpt    3   1.125 ? 0.193  ops/us
UuidBenchmark.randomUuidV4Max     thrpt    3  36.466 ? 4.738  ops/us
UuidBenchmark.randomUuidV4One     thrpt    3   4.546 ? 0.461  ops/us
```

### [`Queue<byte[]>`](https://github.com/palantir/tritium/compare/ds/ids-pool-chunks?expand=1)  , `ConcurrentLinkedDeque`, `ArrayBlockingQueue` & `synchronizedDeque(new ArrayDeque(CHUNKS_TO_ALLOCATE)` are similar result

```
16x Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz cores
VM version: JDK 17.0.8, OpenJDK 64-Bit Server VM, 17.0.8+7-LTS
Benchmark                          Mode  Cnt  Score   Error   Units
UuidBenchmark.jdkRandomUuidMax    thrpt    3  1.680 ? 1.495  ops/us
UuidBenchmark.jdkRandomUuidOne    thrpt    3  2.500 ? 0.073  ops/us
UuidBenchmark.jugSecureRandomMax  thrpt    3  2.003 ? 1.428  ops/us
UuidBenchmark.secureRandomMax     thrpt    3  3.678 ? 0.135  ops/us
UuidBenchmark.threadLocalMax      thrpt    3  2.143 ? 0.564  ops/us
UuidBenchmark.twoLongsMax         thrpt    3  0.723 ? 1.223  ops/us
UuidBenchmark.randomUuidV4Max     thrpt    3  3.531 ? 0.526  ops/us
UuidBenchmark.randomUuidV4One     thrpt    3  2.935 ? 0.135  ops/us
```

```
MacBook Pro 2021 10x aarch64 cores
VM version: JDK 17.0.8.1, OpenJDK 64-Bit Server VM, 17.0.8.1+8-LTS
Benchmark                          Mode  Cnt   Score   Error   Units
UuidBenchmark.jdkRandomUuidMax    thrpt    3   2.305 ? 0.251  ops/us
UuidBenchmark.jdkRandomUuidOne    thrpt    3   3.843 ? 0.098  ops/us
UuidBenchmark.jugSecureRandomMax  thrpt    3   2.952 ? 0.177  ops/us
UuidBenchmark.secureRandomMax     thrpt    3   5.403 ? 0.117  ops/us
UuidBenchmark.threadLocalMax      thrpt    3  10.920 ? 8.761  ops/us
UuidBenchmark.twoLongsMax         thrpt    3   1.153 ? 0.128  ops/us
UuidBenchmark.randomUuidV4Max     thrpt    3   5.402 ? 0.284  ops/us
UuidBenchmark.randomUuidV4One     thrpt    3   4.824 ? 0.228  ops/us
```

**Review by carterkozak — APPROVED** [review]
I appreciate the benchmarks + results you provided with this PR, it really illustrates the upside!

**Pull request merged**